### PR TITLE
Table widget refactors

### PR DESCRIPTION
**Pull request by grahamsmith**
Started looking to see if the Multipage and Table performance can be enhanced with larger datasets.

Thought a small PR to start with would be easier.

Refactored to eliminate nullable doubles and reduce force unboxing.

### CODE REVIEW

**Comment by grahamsmith** [comment]
@DavBfr I think my formatter is different to yours, are you at 80 chars?

File: pdf/lib/src/widgets/table.dart [hunk 1]
```diff
@@ -66,12 +66,13 @@ class TableBorder extends Border {
...
-        bottom: side,
-        left: side,
-        horizontalInside: side,
-        verticalInside: side);
+      top: side,
+      right: side,
+      bottom: side,
+      left: side,
+      horizontalInside: side,
+      verticalInside: side,
```
Comment (grahamsmith) [inline pdf/lib/src/widgets/table.dart:74]: Added trailing comma to force formatting

File: pdf/lib/src/widgets/table.dart [hunk 2]
```diff
@@ -149,15 +149,14 @@ class TableContext extends WidgetContext {
 class ColumnLayout {
   ColumnLayout(this.width, this.flex);
 
-  final double? width;
-  final double? flex;
+  final double width;
```
Comment (grahamsmith) [inline pdf/lib/src/widgets/table.dart:152]: Removing the null here makes the amount the work later on easier. I couldn't find a case where they would be.

File: pdf/lib/src/widgets/table.dart [hunk 3]
```diff
@@ -335,31 +330,29 @@ class Table extends Widget with SpanningWidget {
...
     _heights.clear();
     var index = 0;
 
     for (final row in children) {
-      var n = 0;
-      for (final child in row.children) {
-        final columnWidth = columnWidths != null && columnWidths![n] != null
-            ? columnWidths![n]!
-            : defaultColumnWidth;
+      for (final entry in row.children.asMap().entries) {
```
Comment (grahamsmith) [inline pdf/lib/src/widgets/table.dart:341]: Hopefully this is a bit more legible and no longer has to worry about null as much

**Comment by DavBfr** [comment]
> @DavBfr I think my formatter is different to yours, are you at 80 chars?

Yes I am using the default remanded dart formatting.

**Comment by grahamsmith** [comment]
@DavBfr  - sorted - line lengths are 80 so the diff is simpler now

Unsure about the CI failures from the last run

**Comment by grahamsmith** [comment]
@DavBfr - sorry not relevant to this PR but do not know where else to post as too excited

before my changes for a demo pdf with 10,000 items into a table it took 0:00:57.012834
my new fork with changes 0:00:11.770033

Nothing too major either!

Much excite

**Comment by DavBfr** [comment]
Congrats!

**Comment by grahamsmith** [comment]
I noticed the checks have passed, do I need to do anything else @DavBfr ?

**Comment by DavBfr** [comment]
Just need some time to review.

**Comment by Gustl22** [comment]
@grahamsmith this will potentially conflicting with #1736, where some changes to the algorithm were made. Hope we find a good solution to make it perform good also with cell spans enabled.

**Comment by grahamsmith** [comment]
I imagine it will be terrible, although a potential useful feature.

I have more changes in the background that greatly enhance the performance of the layout as its quite slow as other's have mentioned.

Wonder if the spanning could be in a specific set of widgets. In a perfect world would really love to know how many users need cell spanning (clearly you did).

**Comment by grahamsmith** [comment]
@Gustl22 - do you know when your changes will land? I am not particularly precious about this PR. One of the biggest gains is in MultiPage so I can do a seperate PR for that and hopefully it will not clash.

**Comment by Gustl22** [comment]
> do you know when your changes will land?

No I don't, open source takes time, just wanted to make aware of this conflict. I am also lacking the time to review my stuff in my repos.

> Wonder if the spanning could be in a specific set of widgets. In a perfect world would really love to know how many users need cell spanning (clearly you did).

At least 14 people created or upvoted an issue regarding that. And I think the most people are working around this by creating custom cell tables, but that's not the clean way. I don't mind which one is merged first, I also have to create performance tests.

**Comment by DavBfr** [comment]
Thanks!

**Comment by grahamsmith** [comment]
I did not keep a copy sorry!

From memory I inserted a string with the item count e.g. 'item $index' into a cell on each row of 10000 rows.

Performance tests vary per machine setup too so are usually subjective.

**Comment by Gustl22** [comment]
Thank you, ya just wanted to make this more objective. I repeated the tests multiple times.

**Pull request merged**